Infra: Add actions to publish to ECR & Docker Hub

giom-l opened the pull request:
<!-- ignore-task-list-start -->
**What changes did you make?** (Give an overview)
Add login to dockerhub and push image to it in main workflow
Fixes #237 
Fixes #242

**Is there anything you'd like reviewers to focus on?**
Secrets names may need to be changed based on what is currently available (no visibility on it)

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)
- [x] Yes, on my forked project with own credentials

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code


Check out [Contributing](https://github.com/kafbat/kafka-ui/blob/main/.github/CONTRIBUTING.md) and [Code of Conduct](https://github.com/kafbat/kafka-ui/blob/main/.github/CODE-OF-CONDUCT.md)

![image](https://github.com/kafbat/kafka-ui/assets/29728596/cd0d72b0-016c-4655-be31-63a441ae0e12)

Haarolean reviewed (changes requested):
Hi, could you make this a separate job (within the same workflow) please? So we don't have to rebuild (or re-push) everything if one of the things fail.
Ideally:
- build
- ghcr push
- dockerhub push
- aws ecr push

You'd need to transfer artifacts between the jobs, feel free to use .github/workflows/e2e-*.yml workflows as reference.

giom-l commented:
Hi, 
I spent some time last days to on that and mostly struggled with my own station to have reproducible local builds. 
Anyway, I also took into account https://github.com/kafbat/kafka-ui/issues/349 to go with a solution that would support it.
 
So here are 2 versions that work the same, let me know which one your prefer :D 
In both, I had to activate containerd as builder, for some different reasons (being able to load multi platforms images & preserve provenance attestations) 

With docker save / docker load : 
https://github.com/giom-l/kafka-ui/blob/feat/improve_docker_publish/.github/workflows/save_load.yml

With containerd all along (using `oci` outputs & `ctr`)
https://github.com/giom-l/kafka-ui/blob/feat/improve_docker_publish/.github/workflows/with_ctr.yml

All produced images can be found in all 3 repositories : 
- Github packages: https://github.com/giom-l/kafka-ui/pkgs/container/kafka-ui
- Dockerhub : https://hub.docker.com/repository/docker/gocho/kafka-ui/tags
- ECR public : https://gallery.ecr.aws/j4u0y1h1/kafka-ui?page=1

For ECR, my workflow uses [OIDC provider ](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) by that would be the same with classic credentials (I saw some in your other workflows)

Haarolean commented:
> Hi, I spent some time last days to on that and mostly struggled with my own station to have quicker local builds... Anyway I also took into account #349 to go with a solution that would support it.
> 
> So here are 2 versions that work the same, let me know which one your prefer. In both, I had to activate `containerd` as builder, for some different reasons (being able to load multi platforms images & preserve provenance attestations)
> 
> With docker save / docker load mechanism : https://github.com/giom-l/kafka-ui/blob/feat/improve_docker_publish/.github/workflows/save_load.yml
> 
> With containerd all along (using `oci` output & `ctr` to reuse the produced image) https://github.com/giom-l/kafka-ui/blob/feat/improve_docker_publish/.github/workflows/with_ctr.yml
> 
> All produced images can be found in all 3 repositories :
> 
> * Github packages: https://github.com/giom-l/kafka-ui/pkgs/container/kafka-ui
> * Dockerhub : https://hub.docker.com/repository/docker/gocho/kafka-ui/tags
> * ECR public : https://gallery.ecr.aws/j4u0y1h1/kafka-ui?page=1
> 
> For ECR, my workflow uses [OIDC provider ](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services). It would be the same with classic credentials (I saw some in your other workflows) and even be less verbose by using [this action](https://github.com/jwalton/gh-ecr-push) that would allow something like
> 
> ```
> - name: Push to ECR
>   id: ecr
>   uses: jwalton/gh-ecr-push@v2
>   with:
>     access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
>     secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
>     region: us-east-1
>     local-image: kafka-ui:temp
>     image: <repo>/kafka-ui:main, <repo>/kafka-ui:<sha>
> ```

@giom-l thank you for the analysis. I'd prefer using docker save/load mechanism to have the same approach between different workflows (for example, [e2e-run.yml](https://github.com/kafbat/kafka-ui/blob/main/.github/workflows/e2e-run.yml#L86), already uses this approach).

Can you update the PR branch with this approach so we can proceed with the review? Thank you

giom-l commented:
Sorry for the delay ! 
I updated the PR using the save/load mechanism as requested. 
I also impacted the release workflow that I forgot initially. 

For the ECR publish, it's still "WIP" as I would need some informations about how you want to authenticate and where (ecr public ?)

giom-l commented:
Sure.
Just a question about how you see that :) 
One reusable action per target registry (ecr, github, ecr) ? 
Or one reusable action (something like image-publish) that push to all registries (but I'm not sure how it behave if one of them fails. Can we still rerun just a part of it ?)  ? 

I thought about one generic reusable action that will take parameters (like registry, credentials) but since ECR login is not the same as the other, I'm not sure it's doable. 

Let me know what you think

Haarolean commented:
I see that as follows:
1. A workflow to run on a push to main
2. A workflow to run on a release
3. Both 1) and 2) call another workflow to build an image (docker-build.yml) and a workflow to publish one (docker-publish.yml)
4. docker-publish.yml contains 3 jobs for each registry (gh, ecr, dockerhub)
In this case we can extract common parts for each workflow and keep publishing as separate jobs within a separate calling workflow. 

Feel free to use e2e-*.yml workflows as templates, we have the basic required stuff there (calling other workflows, passing arguments through, transferring artifacts between jobs, etc.)

>Can we still rerun just a part of it 
Yes, we would be able to rerun separate jobs within a workflow like this:
<img width="669" alt="image" src="https://github.com/kafbat/kafka-ui/assets/1494347/3a21fc60-be96-46e3-a4a2-435f7f25485c">

giom-l commented:
Hello

I gave this some time yesterday and made it work as follows (PR udpated) : 
- 1 docker_build.yml workflow
- 1 docker_publish.yml workflow, which itself calls 3 differents workflows : 
     - 1 publish_ecr.yml
     - 1 publish_dockerhub.yml
     - 1 publish_ghcr.yml
     
About extracting common parts for the three publish jobs : 
It seems not really useful to me since it's only a matter of downloading artifacts & configuring docker daemon, which can't be shared between jobs. 

However, I also experienced another solution that can be found in [this branch](https://github.com/giom-l/kafka-ui/blob/feat/matrix_strategy) : 
The main & release job will call the [docker_publish](https://github.com/giom-l/kafka-ui/blob/feat/matrix_strategy/.github/workflows/docker_publish_matrix.yml) and this one is more generic (only the logging part isn't) and uses matrix instead of calling others workflows.
I find this solution to be more elegant (less duplicated code, no intermediary workflow) 

Let me know what you think and I'll finalize the PR with the solution that is preferred on your side.

Haarolean commented on .github/workflows/main.yml:
```diff
@@ -37,42 +41,32 @@ jobs:
-            ${{ runner.os }}-buildx-
+          name: kafbat-ui-${{ steps.build.outputs.version }}
+          path: api/target/api-${{ steps.build.outputs.version }}.jar
+          retention-days: 7
```
let's do 1, the jars are pretty fat to hold in the cache for too long

giom-l replied:
Sure, fixed in [76ed720](https://github.com/giom-l/kafka-ui/commit/76ed720828c409376ba55785e7e4f286cb6ba2aa)

Haarolean commented on .github/workflows/main.yml:
```diff
@@ -9,12 +9,16 @@ permissions:
+
     permissions:
       contents: read
       packages: write
```
I don't think we need packages here anymore as we're building image in a separate job

giom-l replied:
Agreed.

Removed in [76ed720](https://github.com/giom-l/kafka-ui/commit/76ed720828c409376ba55785e7e4f286cb6ba2aa)

Haarolean commented on .github/workflows/publish_dockerhub.yml:
```diff
@@ -0,0 +1,50 @@
+          daemon-config: |
+            {
+              "features": {
+                "containerd-snapshotter": true
```
why is this `containerd-snapshotter` thing required?

giom-l replied:
Without it, SBOM and provenance attestation are not preserved.

I just gave it a test. 

With containerd activated : 

```
docker buildx imagetools inspect ghcr.io/giom-l/kafka-ui:main --format '{{ json .Provenance }}'
{
  "linux/amd64": {
    "SLSA": {
      "buildType": "https://mobyproject.org/buildkit@v1",
      "builder": {
        "id": "https://github.com/giom-l/kafka-ui/actions/runs/9396399214"
      },
      "invocation": {
        "configSource": {
          "entryPoint": "Dockerfile"
        },
        "environment": {
          "platform": "linux/amd64"
        },
        "parameters": {
          "frontend": "dockerfile.v0",
          "locals": [
            {
              "name": "context"
            },
            {
              "name": "dockerfile"
            }
          ]
        }
      },
      "materials": [
        {
          "digest": {
            "sha256": "5ba01ac30fcc3d7be766ad55dae06de90b15fe0d940baac94b23ddb231e8b089"
          },
          "uri": "pkg:docker/azul/zulu-openjdk-alpine@17.0.11-jre-headless?platform=linux%2Famd64"
        },
        {
          "digest": {
            "sha256": "176e0869c38aeaede37e594fcf182c91d44391a932e1d71e99ec204873445a33"
          },
          "uri": "pkg:docker/docker/buildkit-syft-scanner@stable-1"
        }
      ],
      "metadata": {
        "buildFinishedOn": "2024-06-06T06:54:26.509711264Z",
        "buildInvocationID": "pfnymph4d1p9ra92kr3ktm7dv",
        "buildStartedOn": "2024-06-06T06:54:08.308818665Z",
        "completeness": {
          "environment": true,
          "materials": false,
          "parameters": false
        },
        "https://mobyproject.org/buildkit@v1#metadata": {
          "vcs": {
            "localdir:context": "api",
            "localdir:dockerfile": "api",
            "revision": "48d353268657ebb2cc61004f2b59fbe157be8383",
            "source": "https://github.com/giom-l/kafka-ui"
          }
        },
        "reproducible": false
      }
    }
  },
  "linux/arm64": {
    "SLSA": {
      "buildType": "https://mobyproject.org/buildkit@v1",
      "builder": {
        "id": "https://github.com/giom-l/kafka-ui/actions/runs/9396399214"
      },
      "invocation": {
        "configSource": {
          "entryPoint": "Dockerfile"
        },
        "environment": {
          "platform": "linux/amd64"
        },
        "parameters": {
          "frontend": "dockerfile.v0",
          "locals": [
            {
              "name": "context"
            },
            {
              "name": "dockerfile"
            }
          ]
        }
      },
      "materials": [
        {
          "digest": {
            "sha256": "5ba01ac30fcc3d7be766ad55dae06de90b15fe0d940baac94b23ddb231e8b089"
          },
          "uri": "pkg:docker/azul/zulu-openjdk-alpine@17.0.11-jre-headless?platform=linux%2Farm64"
        },
        {
          "digest": {
            "sha256": "176e0869c38aeaede37e594fcf182c91d44391a932e1d71e99ec204873445a33"
          },
          "uri": "pkg:docker/docker/buildkit-syft-scanner@stable-1"
        }
      ],
      "metadata": {
        "buildFinishedOn": "2024-06-06T06:54:26.509711264Z",
        "buildInvocationID": "pfnymph4d1p9ra92kr3ktm7dv",
        "buildStartedOn": "2024-06-06T06:54:08.308818665Z",
        "completeness": {
          "environment": true,
          "materials": false,
          "parameters": false
        },
        "https://mobyproject.org/buildkit@v1#metadata": {
          "vcs": {
            "localdir:context": "api",
            "localdir:dockerfile": "api",
            "revision": "48d353268657ebb2cc61004f2b59fbe157be8383",
            "source": "https://github.com/giom-l/kafka-ui"
          }
        },
        "reproducible": false
      }
    }
  }
}%                                 
```

Exact same workflow, just without containerd

```
docker buildx imagetools inspect ghcr.io/giom-l/kafka-ui:main --format '{{ json .Provenance }}'
{}%                                                                                                                                                
```

Haarolean replied:
thanks!

Haarolean commented:
Hey, sorry for the delay.

One more thing, can we get separate workflows like `publish_x.yml` to be separate jobs within **one** workflow instead? I don't see any point in making them separate workflows rather than jobs, it just adds additional overhead. Let me know what you think.

giom-l commented:
Hello, 

Sure, I'll give some tests to your proposals and fix them if everything's good. 

About 

> One more thing, can we get separate workflows like publish_x.yml to be separate jobs within one workflow instead? I don't see any point in making them separate workflows rather than jobs, it just adds additional overhead. Let me know what you think.

I totally agree. 
This is why I also proposed another solution that I find more elegant. 
You can find the [workflow here](https://github.com/giom-l/kafka-ui/blob/feat/matrix_strategy/.github/workflows/docker_publish_matrix.yml)

It uses matrix strategy to produce as much jobs as we want, but it remains in 1 workflow only (way more readable). 
The downside is about having steps that will run only in certain cases (login to target platform) but it removes duplication of code. 

Let me know if it better fits what you expect and I'll port it to the PR branch :) .

Haarolean commented:
Hi,

the problem with the matrix approach is that's just one job within a workflow which will result in unnecessary re-deployments in case when one of three fails (say, we've deployed to ecr and then ghcr has failed). If we split them in three separate jobs, we won't have to redeploy on job restart. Can we adjust that?

<img width="241" alt="image" src="https://github.com/kafbat/kafka-ui/assets/1494347/69ce6f9c-3705-4763-82a1-4058be572ecf">
<img width="674" alt="image" src="https://github.com/kafbat/kafka-ui/assets/1494347/f3d933ae-bb64-4baf-ac7a-7bb017aff9bf">
<img width="677" alt="image" src="https://github.com/kafbat/kafka-ui/assets/1494347/60c78422-24a3-482c-afdc-4a413c8069cf">

giom-l commented:
Hello, 

This differs from what I'm having on my end. 
I just changed my dockerhub username with wrong value, so the publish failed (obviously)

![image](https://github.com/kafbat/kafka-ui/assets/29728596/d53eb83b-446c-405d-ba44-0181d8812bea)

When I try to rerun all, of course all of them will rerun 
![image](https://github.com/kafbat/kafka-ui/assets/29728596/84d93592-8518-4270-8e46-0116f43aea90)

But when running only failed jobs, only the failed one will rerun, not all of them
![image](https://github.com/kafbat/kafka-ui/assets/29728596/fb574f3a-80f1-4818-8305-97da0ee1feb9)

Also, besides each job, you can rerun them independently 
![image](https://github.com/kafbat/kafka-ui/assets/29728596/7eaf8643-eb69-4271-a047-d55f614d38ac)

But if you confirm you prefer 3 jobs in the same workflow (with duplicate code though, since we can't factorize much), tha'ts fine for me too :)

giom-l commented:
Hello, 

I'm having the same behavior between 3 jobs and a matrix that generates 3 jobs. 
See here, I changed a secret on my project to make the dockerhub publish fail (which happened)

![image](https://github.com/kafbat/kafka-ui/assets/29728596/51be5d9a-6f05-46b2-abdf-daea3a242272)

If I rerun all jobs, all of them will rerun (of course)
![image](https://github.com/kafbat/kafka-ui/assets/29728596/fea638d1-0a2b-453c-9daf-69f042250ee1)


But if I rerun the failed one, only failed one will rerun. 
![image](https://github.com/kafbat/kafka-ui/assets/29728596/5cac2035-ec86-40b3-9660-4c4068d31d3f)

One can also rerun the failed job with the dedicated icon
![image](https://github.com/kafbat/kafka-ui/assets/29728596/d613328c-fb44-4f01-8f04-8b64b831451f)
or after selecting the failed job.
![image](https://github.com/kafbat/kafka-ui/assets/29728596/0dcd1eec-6d22-4d3f-83c7-559a82faeeb2)


And you can see here that the previously succeeded jobs didn't rerun when the failed one restart
![image](https://github.com/kafbat/kafka-ui/assets/29728596/c6ad1ac9-b87e-4e3e-b9ce-c4811d9bc088)

Haarolean commented on .github/workflows/publish_ghcr.yml:
```diff
@@ -0,0 +1,60 @@
+
+      - name: Push images to GHCR
+        run: |
+          # docker tag kafka-ui:temp ghcr.io/kafbat/kafka-ui:${{ inputs.generic_tag }}
```
can we clean this up please?
also, I guess we could use github repo name here to get rid of the hardcoded string

giom-l replied:
Workflow removed since we switched to matrix.

Haarolean commented on .github/workflows/publish_ecr.yml:
```diff
@@ -0,0 +1,64 @@
+      - name: Push to ECR
+        env:
+          REGISTRY: ${{steps.login-ecr-public.outputs.registry }}
+          REGISTRY_ALIAS: j4u0y1h1
```
is this needed? is it like an ecr id?

giom-l replied:
Nope, that's my ECR repo id :D 
This workflow has been removed so it can be closed.

Haarolean commented:
>I'm having the same behavior between 3 jobs and a matrix that generates 3 jobs.

Oh, thanks for correcting me! I've seen matrix workflows in action but haven't used them personally before so I thought the part we need was missing :) 

Besides the minor comments I've left (and other possible things to clean up you manage to find) I'd need to get our credentials for ECR and we can merge. Thank you very much for the high-quality PR and continuous cooperation, and I am sorry for the delays on my part :)

giom-l commented:
Hello, 

I just updated the PR to use matrix workflows and removed old workflows that contained some hardcoded strings.

It should work with no issue as soon as : 
- The secrets used in the workflows are correct (let me know if I need to change some names)
```
secrets.DOCKERHUB_USERNAME
secrets.DOCKERHUB_TOKEN
```

- You are using ECR public and not private.
- AWS authentication is handled with an AWS role 
```
secrets.AWS_ROLE
```
This is a personal preference to use role over static credentials.
If you prefer to stick with credentials, it will be easy to change. Let me know.

- The owner/repository names are the same for all 3 registries (`kafbat/kafka-ui`)
If that is not the case, we'll need to introduce another variable "OWNER" in the `docker_publish.yml` workflow to take care of this difference.

Haarolean commented on .github/workflows/docker_build.yml:
```diff
@@ -0,0 +1,88 @@
+
+permissions:
+  contents: read
+  packages: write
```
Do we need this here? As we're only building here and not writing packages

giom-l replied:
Oh, you're right. 
Totally useless. 
Fixed in f2a65045

Haarolean commented on .github/workflows/docker_publish.yml:
```diff
@@ -0,0 +1,97 @@
+        required: true
+        type: string        
+
+jobs:
```
This workflow, on the contrary, might need packages:write for ghcr publish?

Haarolean replied:
Or permissions: inherit?

giom-l replied:
That would be really nice to also have inheritance for permissions but I didn't find anything like this. 
However, I added 
```
permissions: 
  packages: write
```
in f2a65045

and also align those permissions in `main.yml` and `release.yml` workflows

Haarolean replied:
Thanks much! 
> I didn't find anything like this in documentation.

I must've mixed this with secrets :) 

I'm currently waiting to get the secrets configured, we'll merge once I get that done

Haarolean commented:
@giom-l thanks for your first (and a high quality one!) contribution to kafbat UI! I'll be taking a look at how this behaves now.

Haarolean commented:
@giom-l any clues on what's the unknown/unknown image arch we've got? 
https://github.com/kafbat/kafka-ui/pkgs/container/kafka-ui/285510516?tag=07f0e0e75471a44be52782b06d0362c42820cba1
<img width="746" alt="image" src="https://github.com/user-attachments/assets/a05c3fe2-2b5a-4a61-9fd4-d2bd7730059d">

giom-l commented:
It reminds me of something about provenance but can't remember exactly what  ATM. 

I just found this discussion that seem related : 
https://github.com/orgs/community/discussions/45969

But there does not seem to be proper answer (get rid of provenance, pinning version that seem to not work anymore...)
